src/Evolution/Systems/Burgers/Observe.hpp

```diff
@@ -4,141 +4,170 @@
+#include "IO/Observer/Helpers.hpp"  // IWYU pragma: keep
 #include "IO/Observer/ObservationId.hpp"
-#include "IO/Observer/ObserverComponent.hpp"
-#include "IO/Observer/ReductionActions.hpp"
-#include "IO/Observer/VolumeActions.hpp"
+#include "IO/Observer/ObserverComponent.hpp"  // IWYU pragma: keep
+#include "IO/Observer/ReductionActions.hpp"  // IWYU pragma: keep
+#include "IO/Observer/Tags.hpp"
+#include "IO/Observer/VolumeActions.hpp"  // IWYU pragma: keep
 #include "Options/Options.hpp"
+#include "Parallel/ArrayIndex.hpp"
+#include "Parallel/CharmPupable.hpp"
 #include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Reduction.hpp"
 #include "PointwiseFunctions/AnalyticSolutions/Tags.hpp"
-#include "Time/Tags.hpp"
+#include "Time/Time.hpp"
 #include "Utilities/Functional.hpp"
 #include "Utilities/MakeString.hpp"
 #include "Utilities/Numeric.hpp"
+#include "Utilities/Registration.hpp"
 #include "Utilities/TMPL.hpp"
 #include "Utilities/TaggedTuple.hpp"
 
-namespace Burgers {
-namespace Actions {
+/// \cond
+// IWYU pragma: no_forward_declare Tensor
+namespace Tags {
+struct Time;
+}  // namespace Tags
+/// \endcond
 
-namespace observe_detail {
-using reduction_datum = Parallel::ReductionDatum<double, funcl::Plus<>,
-                                                 funcl::Sqrt<funcl::Divides<>>,
-                                                 std::index_sequence<1>>;
-using reduction_datums =
-    tmpl::list<Parallel::ReductionDatum<double, funcl::AssertEqual<>>,
-               Parallel::ReductionDatum<size_t, funcl::Plus<>>,
-               reduction_datum>;
+namespace Burgers {
+namespace Events {
+template <typename EventRegistrars>
+class Observe;
 
-}  // namespace observe_detail
+namespace Registrars {
+using Observe = Registration::Registrar<Events::Observe>;
+}  // namespace Registrars
 
 /*!
- * \brief Temporary action for observing volume and reduction data
+ * \brief Observe the Burgers system
  *
- * A few notes:
- * - Writes the solution and error in \f$U\f$ to disk.
- * - The RMS error of \f$U\f$ is written to disk.
+ * Writes volume quantities:
+ * - `U`
+ * - `ErrorU` = \f$U - \text{analytic solution}\f$
+ *
+ * Writes reduction quantities:
+ * - `ErrorU` = \f$\operatorname{RMS}(U - \text{analytic solution})\f$
+ *   over all points
+ *
+ * \warning Currently, only one observation event can be triggered at
+ * a given time.  Causing multiple events to run at once will produce
+ * unpredictable results.
  */
-struct Observe {
-  struct ObserveNSlabs {
-    using type = size_t;
-    static constexpr OptionString help = {"Observe every Nth slab"};
-  };
-  struct ObserveAtT0 {
-    using type = bool;
-    static constexpr OptionString help = {"If true observe at t=0"};
-  };
-
-  using const_global_cache_tags = tmpl::list<ObserveNSlabs, ObserveAtT0>;
+template <typename EventRegistrars = tmpl::list<Registrars::Observe>>
+class Observe : public Event<EventRegistrars> {
```

Comment:
If I understand correctly, we are going to be able to use this event to observe e.g. at a particular time, at every N steps, or at the final time. Therefore I think it makes more sense to split this action into _two_ events: One for reductions and one for the volume. Then we can observe reductions more frequently than the volume, for instance.

Reply:
Agreed.  That is actually one of the later PRs in this series.

The rough sequence I have planned is (not entirely sure how it will be split up):
1. Event for Burgers (this PR)
2. Event for arbitrary system
3. Split volume and reduction events

I may also deal with observing arbitrary fields, (as opposed to just the system variables) but it is possible that will get put off.

Reply:
nice 👍 sounds like less work to split the events straight away though

Reply:
Perhaps, but I'm trying to be kind to my reviewers.

Reply:
I'd prefer to review the "final" (split) version, instead of the intermediate step. @kidder how about you?

Reply:
Talking with @wthrowe , it sounds like this PR is essentially finished, and the follow-ups are almost ready (2 and 3 above) so I think it is less effort for @wthrowe to just merge this now.

Reply:
It seems to me that splitting this and generalizing to an arbitrary system would revise almost all changes of the “Convert to event” commit. I don’t see the point of doing that, especially if you have these changes ready. If you don’t want to rewrite the commit order I’d be fine with just pushing them to this PR and later squash them. Then we can review the full picture, instead of code that’s going to change anyway.

Reply:
I don't want that, I like that @wthrowe is feeding it in as smaller logical pieces. I really don't see the benefit of turning this into a monster 2000+ line PR. We're going ahead with the order @wthrowe has planned.